### PR TITLE
Fix ref counting when using futures in AbstractClient

### PR DESCRIPTION
**Pull request by original-brownbear**
Found this now that #102030 is getting closer to completion. It's fundamentally broken how the client deals with ref counted messages. We were only saved by the fact that currently we do do not handle any messages with non-noop ref counting through this client interface. 

We fundamentally need to increment the ref count by one before assigning a ref counted value to a future result.
Otherwise, transport actions will have to understand the specific kind of listener they resolve and cannot decrement sent/consumed transport messages themselves.

marking non-issue since this hasn't caused any production trouble yet.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/client/internal/OriginSettingClientTests.java [hunk 1]
```diff
@@ -37,7 +37,6 @@ protected <Request extends ActionRequest, Response extends ActionResponse> void
                     ActionListener<Response> listener
                 ) {
                     assertEquals(origin, threadPool().getThreadContext().getTransient(ThreadContext.ACTION_ORIGIN_TRANSIENT_NAME));
-                    super.doExecute(action, request, listener);
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/client/internal/OriginSettingClientTests.java:40]: Neither here nor in the other test do we care about resolving the listener. The result passed by the noop client is always `null`. I figured this was still better than adding a null check or hackily casting a non-null response in the  noop-client ...

File: server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java [hunk 2]
```diff
@@ -448,4 +449,18 @@ public static <T, E extends Exception> T get(CheckedConsumer<PlainActionFuture<T
...
         return fut.actionGet(timeout, unit);
     }
+
+    /**
+     * Same as {@code PlainActionFuture} but for use with {@link RefCounted} result types. Unlike {@code PlainActionFuture} this future
+     * acquires a reference to its result. This means that the result reference must be released by a call to {@link RefCounted#decRef()}
+     * on the result before it goes out of scope.
+     * @param <R> reference counted result type
+     */
+    public static class ForRefCounted<R extends RefCounted> extends PlainActionFuture<R> {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java:459]: We will probably need something similar for listenable future as well at some point. I didn't add any tests here yet before we're cool with the approach, I can though obviously it's trivial :)

**Review by DaveCTurner — COMMENTED** [review]
Hmm I see the problem but this is a fairly awful hack IMO. That said I don't see great alternatives without somewhat invasive changes. The whole concept of blocking calls to the client is kinda bad, but does have production users (mostly Watcher and Monitoring it seems).

For a general-purpose utility I think the whole future should implement `Releasable` so that there's a proper `incRef`/`decRef` pair to keep things clear. As written here, there's an implicit assumption that there's only one call to `.get()` the result: if there are multiple `.get()` callers then they would need to coordinate the single required `decRef()` call amongst themselves. That seems super-trappy to me.

Also for a general-purpose utility we'd need to handle the case where the future is completed multiple times, in which we only `incRef` the result if it's the first completion.

Maybe it'd be better to make this specific behaviour local to the client rather than a general-purpose utility. I expect we've more chance of being able to assert that there's only a single call to `.get()` the result in that case. 

Maybe also it should be a different API on the client for those few actions where refs matter so that it's clear to the caller they are taking a ref when calling `.get()`. Not 100% sure about that tho.

**Comment by original-brownbear** [comment]
@DaveCTurner I admit it, it's hacky :) How about this, I moved the custom future into the client to limit the blast radius and I promise I'll look into removing the production usage of blocking get?
I'd mostly like to get this through quickly and without a lot of effort because:
1. the blocking usage in prod must go away anyway
2. this is pretty urgent because it blocks the search ref counting work :)

File: server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java [hunk 3]
```diff
@@ -1598,4 +1599,18 @@ protected <Request extends ActionRequest, Response extends ActionResponse> void
...
+     * Same as {@link PlainActionFuture} but for use with {@link RefCounted} result types. Unlike {@code PlainActionFuture} this future
+     * acquires a reference to its result. This means that the result reference must be released by a call to {@link RefCounted#decRef()}
+     * on the result before it goes out of scope.
+     * @param <R> reference counted result type
+     */
+    private static class RefCountedFuture<R extends RefCounted> extends PlainActionFuture<R> {
+        @Override
+        public final void onResponse(R result) {
+            result.incRef();
+            super.onResponse(result);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java:1613]: Maybe this'd be a little safer?

```suggestion
            assert result.hasReferences();
            if (set(result)) {
                result.incRef();
            }
```

**Review by DaveCTurner — COMMENTED** [review]
Yep I can live with this for now to get the search response work through. I left a suggestion for a slightly safer implementation. Could you also add something to assert that `.get()` is only called once? And we need to do something with the timed `.get()` because a timeout would be a leak, although I'm not sure what exactly would be best there.

**Comment by DaveCTurner** [comment]
Suggestion for a more general-purpose solution: https://github.com/elastic/elasticsearch/pull/102507

**Comment by original-brownbear** [comment]
@DaveCTurner I'll have a look later :) But for now, wdyt about this solution? I made it safer with your suggestion and added an assertion about this only getting called once. Good enough to unblock this one? :)

**Review by DaveCTurner — COMMENTED** [review]
Maybe I'm misreading something but it looks like you've got both impls now?

image

**Comment by original-brownbear** [comment]
I'm so sorry ... 🤦 https://github.com/elastic/elasticsearch/pull/102498/commits/cd4fdd5d39ecaa52b9d9143d4912aeab1d4fa2a5

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
https://github.com/elastic/elasticsearch/pull/102515 would have saved some of the mockery muckery I think

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**